crates/engine/src/ext.rs

```diff
@@ -0,0 +1,414 @@
+// Copyright 2018-2021 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Provides the same interface as Substrate's FRAME `contract` module.
+//!
+//! See [the documentation for the `contract` module](https://docs.rs/crate/pallet-contracts)
+//! for more information.
+
+use crate::{
+    exec_context::ExecContext,
+    storage::Storage,
+    test_api::{
+        self,
+        EmittedEvent,
+        RecInstance,
+    },
+    types::{
+        AccountId,
+        Balance,
+        Key,
+    },
+    OnInstance,
+};
+
+use core::cell::RefCell;
+use std::collections::HashMap;
+
+type Result = core::result::Result<(), Error>;
+
+pub struct EnvInstance {
+    /// The environment storage.
+    pub storage: Storage<Key, Vec<u8>>,
+    /// Holds the balance of each account.
+    pub balances: HashMap<AccountId, Balance>,
```

Comment:
I view it this way: If the store-using-a-prefix implementation is correct there should be no difference from having a separate balance-HashMap as I currently have. Having this separate balance-HashMap makes the implementation a lot easier though. Hence I would opt for keeping it simple like it is now.

Reply:
The fundamental question is: How far do we want to go with the emulation of the contract pallet. Since the old off-chain engine suffered from being too far away from the contracts pallet due to design considerations like exactly these I think it is a very good idea to keep being as close as possible.
This for example means that users (as with Substrate) can kinda predict where in the storage certain fields are going to be stored and therefore use chain extensions to interopt. with chain state. This is not possible if we model balances in a unique way from how it is done in Substrate. Maybe this is not a problem now since we do not yet support all the different use cases in the current MVP but it will pop up.